runtime/ts/particle-execution-context.ts

```diff
@@ -103,24 +103,47 @@ export class ParticleExecutionContext {
         readonly particle: ParticleSpec;
         private handlers = new Map<string, ((event: {}) => void)[]>();
         private pec: ParticleExecutionContext;
+        private providedSlots: Map<string, string>;
         private requestedContentTypes = new Set<string>();
         private _isRendered = false;
-        constructor(pec: ParticleExecutionContext, particle: ParticleSpec, slotName: string) {
+        constructor(pec: ParticleExecutionContext, particle: ParticleSpec, slotName: string, providedSlots: Map<string, string>) {
           this.slotName = slotName;
           this.particle = particle;
           this.pec = pec;
+          this.providedSlots = providedSlots;
         }
         get isRendered() { return this._isRendered; }
         /** @method render(content)
          * renders content to the slot.
          */
         render(content) {
+          if (content.template && this.providedSlots.size > 0) {
+            content = {...content};
+            if (typeof content.template === 'string') {
+              content.template = this._substituteSlotNamesForIds(content.template);
+            } else {
+              content.template = Object.entries(content.template).reduce(
+                  (templateDictionary, [templateName, templateValue]) => {
+                    templateDictionary[templateName] = this._substituteSlotNamesForIds(templateValue);
+                    return templateDictionary;
+                  }, {});
+            }
+          }
+
           this.pec.apiPort.Render({particle, slotName, content});
 
           Object.keys(content).forEach(key => { this.requestedContentTypes.delete(key); });
           // Slot is considered rendered, if a non-empty content was sent and all requested content types were fullfilled.
           this._isRendered = this.requestedContentTypes.size === 0 && (Object.keys(content).length > 0);
         }
+        _substituteSlotNamesForIds(template) {
```

Comment:
best practice in .ts files is to mark private methods as private rather than using an underscore.

Reply:
Thanks for pointing out. I need to refresh my memory of TS capabilities and syntax.